### PR TITLE
Audio: Implement a HiFi3 version audio_stream_copy

### PR DESCRIPTION
**Pull request by andrula-song**
Since Xtensa provides intrinsics for multi-way load/store and using those instructions can reduce more cycles than the function we using now, so there is a HiFi3 audio_stream _copy implementation.

Signed-off-by: Andrula Song <xiaoyuan.song@intel.com>

### CODE REVIEW

**Comment by andrula-song** [comment]
Function audio_stream _copy  is used by many components like copier, eq, kpb,drc,src,  etc. So any improvement of it is meaningful, eventhough the percentage is not very huge. Test with my TGL Andrews device, the audio_stream _copy function itself can save about 12% cycles then the original one. and from the perspective of component copier, it can save about 3.5% cycles.  Here are some logs from mtrace:
the original version log:
"
[  107.958665] <inf> copier: comp:1 0x40001 do_conversion_copy gateway on output cycles is 112
[  108.958665] <inf> copier: comp:1 0x40001 do_conversion_copy gateway on output cycles is 112

[    0.434460] <inf> component: comp:1 0x40001 perf comp_copy samples 0 period 1000 cpu avg 756 peak 767
[    1.458460] <inf> component: comp:1 0x40001 perf comp_copy samples 0 period 1000 cpu avg 756 peak 767
"
the optimized version log:
"
[  107.546456] <inf> copier: comp:1 0x40001 do_conversion_copy gateway on output cycles is 98
[  108.546456] <inf> copier: comp:1 0x40001 do_conversion_copy gateway on output cycles is 98

[    4.020263] <inf> component: comp:1 0x40001 perf comp_copy samples 0 period 1000 cpu avg 727 peak 747
[    5.044263] <inf> component: comp:1 0x40001 perf comp_copy samples 0 period 1000 cpu avg 727 peak 747
"

**Review by lgirdwood — COMMENTED** [review]
LGTM, @singalsu pls review.

File: src/audio/component.c [hunk 1]
```diff
@@ -169,6 +176,53 @@ void comp_get_copy_limits_frame_aligned(const struct comp_buffer __sparse_cache
...
+	/* copy with 16bit as the minimum unit since the minimum sample size is 16 bit*/
+	while (shorts) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		shorts_src = audio_stream_samples_without_wrap_s16(source, src);
+		shorts_dst = audio_stream_samples_without_wrap_s16(sink, dst);
+		shorts_copied = AE_MIN_32_signed(shorts_src, shorts_dst);
+		shorts_copied = AE_MIN_32_signed(shorts, shorts_copied);
+		m = shorts_copied >> 2;
+		left = shorts_copied & 0x03;
```
Comment (btian1) [inline src/audio/component.c:205]: & also can be achieved with % 4

File: src/audio/component.c [hunk 2]
```diff
@@ -169,6 +176,53 @@ void comp_get_copy_limits_frame_aligned(const struct comp_buffer __sparse_cache
...
+		}
+		AE_SA64POS_FP(outu, dst);
+
+		/* process the left bits that less than 4 * 16 */
+		for (i = 0; i < left ; i++) {
+			AE_L16_IP(in_sample, (ae_int16 *)src, sizeof(ae_int16));
+			AE_S16_0_IP(in_sample, (ae_int16 *)dst, sizeof(ae_int16));
+		}
+		shorts -= shorts_copied;
+	}
```
Comment (btian1) [inline src/audio/component.c:220]: wonder still need while()? if need while, then how to explain the left part loop? line 215 - 218?

File: src/audio/component.c (hunk 1 above)
Comment (andrula-song) [reply src/audio/component.c:205]: & faster than %

File: src/audio/component.c (hunk 2 above)
Comment (andrula-song) [reply src/audio/component.c:220]: left part is for those shorts that less than 4 and can not processed by AE_LA16X4_IP and AE_SA16X4_IP

**Review by singalsu — APPROVED** [review]

**Comment by lgirdwood** [comment]
@andrula-song any ETA for non draft ?

**Comment by andrula-song** [comment]
hi @lgirdwood , is there anything needs more modification?

**Review by kv2019i — APPROVED** [review]

**Pull request merged**